src/handshake/server.rs

```diff
@@ -83,15 +86,35 @@ impl<'a, T: AsyncRead + AsyncWrite + Unpin> Server<'a, T> {
     }
 
     /// Await an incoming client handshake request.
-    pub async fn receive_request(&mut self) -> Result<ClientRequest<'a>, Error> {
+    pub async fn receive_request(&mut self) -> Result<ClientRequest<'_>, Error> {
         self.buffer.clear();
+
+        let mut skip = 0;
+
         loop {
             crate::read(&mut self.socket, &mut self.buffer, BLOCK_SIZE).await?;
-            if let Parsing::Done { value, offset } = self.decode_request()? {
-                self.buffer.advance(offset);
-                return Ok(value)
+
+            let limit = std::cmp::min(self.buffer.len(), MAX_HEADERS_SIZE);
+
+            // We don't expect body, so can search for the CRLF headers tail from
+            // the end of the buffer.
+            if self.buffer[skip..limit].windows(4).rev().any(|w| w == b"\r\n\r\n") {
+                break;
+            }
+
+            // Give up if we've reached the limit. We could emit a specific error here,
+            // but httparse will produce meaningful error for us regardless.
+            if limit == MAX_HEADERS_SIZE {
```

Comment:
ok, this will also emit an error if `self.buffer.len() > MAX_HEADER_SIZE`?

it's truncated below by `std::cmp::min` a couple lines above